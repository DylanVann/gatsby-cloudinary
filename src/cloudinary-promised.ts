const cloudinary = require('cloudinary')
import { urlExists } from 'url-exists-promise'
import fs from 'fs'
import axios from 'axios'

const videoExtensions = ['mp4']
const imageExtensions = ['png', 'jpg']

export interface CloudinaryOptions {
    cloud_name: string
    api_key: string
    api_secret: string
}

export const isVideo = (localAbsolutePath: string) => {
    for (let i = 0; i < videoExtensions.length; i++) {
        const extension = videoExtensions[i]
        if (localAbsolutePath.endsWith(extension)) return true
    }
    return false
}

export const isImage = (localAbsolutePath: string) => {
    for (let i = 0; i < imageExtensions.length; i++) {
        const extension = imageExtensions[i]
        if (localAbsolutePath.endsWith(extension)) return true
    }
    return false
}

export const uploadFile = (
    id: string,
    localAbsolutePath: string,
    options: CloudinaryOptions,
): Promise<any> =>
    axios
        .post(
            `https://api.cloudinary.com/v1_1/${options.cloud_name}/${
                isVideo(localAbsolutePath) ? 'video' : 'image'
            }>/upload`,
            fs.createReadStream(localAbsolutePath),
            {
                params: {
                    public_id: id,
                    api_key: options.api_key,
                    api_secret: options.api_secret,
                },
            },
        )
        .then(({ data }) => data)

export const getMetadata = (
    id: string,
    localAbsolutePath: string,
    options: CloudinaryOptions,
): Promise<any> =>
    axios.get(
        `https://api.cloudinary.com/v1_1/${options.cloud_name}/${
            isVideo(localAbsolutePath) ? 'video' : 'image'
        }>/upload`,
        {
            params: {
                public_id: id,
                api_key: options.api_key,
                api_secret: options.api_secret,
                image_metadata: true,
            },
        },
    )

export const imageExists = (id: string, options: CloudinaryOptions): Promise<boolean> => {
    const urlImg = `http://res.cloudinary.com/${options.cloud_name}/image/upload/${id}`
    return urlExists(urlImg)
}

export const videoExists = (id: string, options: CloudinaryOptions): Promise<boolean> => {
    const urlVideo = `http://res.cloudinary.com/${options.cloud_name}/video/upload/${id}`
    return urlExists(urlVideo)
}

export const fileExists = (
    id: string,
    localAbsolutePath: string,
    options: CloudinaryOptions,
): Promise<boolean> => {
    if (isVideo(localAbsolutePath)) {
        return videoExists(id, options)
    }
    return imageExists(id, options)
}

export const uploadOrGetMetadata = async (
    id: string,
    localAbsolutePath: string,
    options: CloudinaryOptions,
): Promise<any> => {
    const exists = await fileExists(id, localAbsolutePath, options)
    if (!exists) {
        // Have to upload the image or video
        return uploadFile(id, localAbsolutePath, options)
    } else {
        // Already uploaded, we just get the metadata
        return getMetadata(id, localAbsolutePath, options)
    }
}
