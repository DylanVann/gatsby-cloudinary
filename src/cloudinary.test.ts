import 'dotenv/config'
import { uploadFile } from './cloudinary-promised'
import * as path from 'path'

const cloudName: string = (process.env.CLOUDINARY_CLOUD_NAME as any) as string
const apiSecret: string = (process.env.CLOUDINARY_API_SECRET as any) as string
const apiKey: string = (process.env.CLOUDINARY_API_KEY as any) as string

it('uploads image', async () => {
    await uploadFile('cloudinary', path.join(__dirname, './cloudinary.png'), {
        cloud_name: cloudName,
        api_secret: apiSecret,
        api_key: apiKey,
    })
})

it('fails to upload large image', async () => {
    await expect(
        uploadFile('cloudinary', path.join(__dirname, './cloudinary-large.png'), {
            cloud_name: cloudName,
            api_secret: apiSecret,
            api_key: apiKey,
        }),
    ).rejects.toThrowErrorMatchingSnapshot()
})
