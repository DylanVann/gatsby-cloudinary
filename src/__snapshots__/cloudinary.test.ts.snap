// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails to upload large image 1`] = `"File size too large. Got 10574322. Maximum is 10485760."`;
